File: CS/CustomBandedGridView/CustomBandedGridView.cs
using DevExpress.Utils;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Views.BandedGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Base.Handler;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Printing;
using DevExpress.XtraPrinting;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

namespace dxExample {
    public class CustomBandedGridView : BandedGridView {
        public const string SViewName = "CustomBandedView";
        protected override string ViewName { get { return SViewName; } }
        public CustomBandedGridView(GridControl ownerGrid)
            : base(ownerGrid) {
                HeaderHeightOnExport = 50;
        }
        public CustomBandedGridView() {

        }
        public int HeaderHeightOnExport { get; set; }
        protected override GridColumnCollection CreateColumnCollection() {
            return new MyBandedColumnCollection(this);
        }
        protected override BaseViewPrintInfo CreatePrintInfoInstance(PrintInfoArgs args) {
            return new CustomBandedPrintInfo(args);
        }

    }
    public class CustomBandedPrintInfo : BandedGridViewPrintInfo {
        BrickStyle vert;
        public CustomBandedPrintInfo(PrintInfoArgs args)
            : base(args) {

        }
        protected override void SetDefaultBrickStyle(IBrickGraphics graph, BrickStyle style) {
            base.SetDefaultBrickStyle(graph, style);
        }

        protected override void PrintBandHeader(IBrickGraphics graph) {
            if(!View.OptionsPrint.PrintBandHeader) return;
            Rectangle r = Rectangle.Empty;
            Point indent = new Point(Indent, 0);
            bool usePrintStyles = View.OptionsPrint.UsePrintStyles;
            SetDefaultBrickStyle(graph, (BrickStyle)Bricks["BandPanel"].Clone());
            foreach(PrintBandInfo band in Bands) {
                r = band.Bounds;
                r.Offset(indent);
                if(!usePrintStyles && band.Band != null) {
                    AppearanceObject temp = new AppearanceObject();
                    AppearanceHelper.Combine(temp, new AppearanceObject[] { band.Band.AppearanceHeader, AppearancePrint.BandPanel });
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1));
                }
                var brick = DrawTextBrick(graph, band.Band.GetTextCaption(), r);
                brick.Style = (BrickStyle)brick.Style.Clone();
                brick.Style.StringFormat.ChangeFormatFlags(StringFormatFlags.NoWrap | StringFormatFlags.LineLimit);


            }
        }
        public override void PrintHeader(IBrickGraphics graph) {
            vert = new BrickStyle(Bricks["HeaderPanel"]);
            vert.StringFormat = vert.StringFormat.ChangeFormatFlags(StringFormatFlags.DirectionVertical | StringFormatFlags.DirectionRightToLeft);
            base.PrintHeader(graph);
            if(!View.OptionsPrint.PrintHeader) return;
            Point indent = new Point(Indent, HeaderY);
            Rectangle r = Rectangle.Empty;
            bool usePrintStyles = View.OptionsPrint.UsePrintStyles;
            SetDefaultBrickStyle(graph, Bricks["HeaderPanel"]);

            int HeaderHeight = (this.View as CustomBandedGridView).HeaderHeightOnExport;
          
            Dictionary<PrintColumnInfo, ITextBrick> columnBricks = new Dictionary<PrintColumnInfo, ITextBrick>();
            foreach(PrintColumnInfo col in Columns) {
                if(!usePrintStyles) {
                    AppearanceObject temp = new AppearanceObject();
                    AppearanceHelper.Combine(temp, new AppearanceObject[] { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel });
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1));
                }
                r = col.Bounds;
                r = new Rectangle(r.Location, new Size(r.Width, HeaderHeight));
                r.Offset(indent);
                MethodInfo mi = typeof(GridColumn).GetMethod("GetTextCaptionForPrinting", BindingFlags.NonPublic | BindingFlags.Instance);
                string caption = (string)mi.Invoke(col.Column, new object[] { });//col.Column.GetTextCaptionForPrinting();
                if(!col.Column.OptionsColumn.ShowCaption) caption = string.Empty;
                ITextBrick itb = DrawTextBrick(graph, caption, r);
                if((col.Column is CustomBandedColumn) && ((col.Column as CustomBandedColumn).CaptionVerticalOnExport))
                    itb.Style = vert;
                else {
                    SetDefaultBrickStyle(graph, Bricks["HeaderPanel"]);
                }
                if(caption.Contains(Environment.NewLine))
                    itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, true);

                if(AppearancePrint.HeaderPanel.TextOptions.WordWrap == WordWrap.NoWrap && View.OptionsPrint.UsePrintStyles) {
                    using(Graphics g = this.View.GridControl.CreateGraphics()) {
                        SizeF s = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value);
                        if(s.Width + 5 >= r.Width) {
                            itb.Text = "";
                            itb.TextValue = "";
                        }
                    }
                }
                columnBricks.Add(col, itb);
                using(Graphics g = this.View.GridControl.CreateGraphics()) {
                    SizeF s = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value);
                    if(s.Height + 5 >= r.Height) {
                        itb.Text = "";
                        itb.TextValue = "";
                    }
                }
            }
        }
    }

    class MyBandedColumnCollection : BandedGridColumnCollection {
        public MyBandedColumnCollection(ColumnView view)
            : base(view) {

        }
        protected override GridColumn CreateColumn() {
            return new CustomBandedColumn();
        }

    }
    class CustomBandedColumn : BandedGridColumn {
        public CustomBandedColumn() {
            CaptionVerticalOnExport = false;
        }
        protected override void Assign(GridColumn column) {
            if(column is CustomBandedColumn)
                this.CaptionVerticalOnExport = ((CustomBandedColumn)column).CaptionVerticalOnExport;
            base.Assign(column);
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        public bool CaptionVerticalOnExport { get; set; }
    }
}

File: VB/CustomBandedGridView/CustomBandedGridView.vb
Imports DevExpress.Utils
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Views.BandedGrid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Base.Handler
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Printing
Imports DevExpress.XtraPrinting
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Windows.Forms

Namespace dxExample
    Public Class CustomBandedGridView
        Inherits BandedGridView

        Public Const SViewName As String = "CustomBandedView"
        Protected Overrides ReadOnly Property ViewName() As String
            Get
                Return SViewName
            End Get
        End Property
        Public Sub New(ByVal ownerGrid As GridControl)
            MyBase.New(ownerGrid)
                HeaderHeightOnExport = 50
        End Sub
        Public Sub New()

        End Sub
        Public Property HeaderHeightOnExport() As Integer
        Protected Overrides Function CreateColumnCollection() As GridColumnCollection
            Return New MyBandedColumnCollection(Me)
        End Function
        Protected Overrides Function CreatePrintInfoInstance(ByVal args As PrintInfoArgs) As BaseViewPrintInfo
            Return New CustomBandedPrintInfo(args)
        End Function

    End Class
    Public Class CustomBandedPrintInfo
        Inherits BandedGridViewPrintInfo

        Private vert As BrickStyle
        Public Sub New(ByVal args As PrintInfoArgs)
            MyBase.New(args)

        End Sub
        Protected Overrides Sub SetDefaultBrickStyle(ByVal graph As BrickGraphics, ByVal style As BrickStyle)
            MyBase.SetDefaultBrickStyle(graph, style)
        End Sub

        Protected Overrides Sub PrintBandHeader(ByVal graph As BrickGraphics)
            If Not View.OptionsPrint.PrintBandHeader Then
                Return
            End If
            Dim r As Rectangle = Rectangle.Empty
            Dim indent As New Point(Me.Indent, 0)
            Dim usePrintStyles As Boolean = View.OptionsPrint.UsePrintStyles
            SetDefaultBrickStyle(graph, DirectCast(Bricks("BandPanel").Clone(), BrickStyle))
            For Each band As PrintBandInfo In Bands
                r = band.Bounds
                r.Offset(indent)
                If (Not usePrintStyles) AndAlso band.Band IsNot Nothing Then
                    Dim temp As New AppearanceObject()
                    AppearanceHelper.Combine(temp, New AppearanceObject() { band.Band.AppearanceHeader, AppearancePrint.BandPanel })
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1))
                End If
                Dim brick = DrawTextBrick(graph, band.Band.GetTextCaption(), r)
                brick.Style = DirectCast(brick.Style.Clone(), BrickStyle)
                brick.Style.StringFormat.ChangeFormatFlags(StringFormatFlags.NoWrap Or StringFormatFlags.LineLimit)


            Next band
        End Sub
        Public Overrides Sub PrintHeader(ByVal graph As BrickGraphics)
            vert = New BrickStyle(Bricks("HeaderPanel"))
            vert.StringFormat = vert.StringFormat.ChangeFormatFlags(StringFormatFlags.DirectionVertical Or StringFormatFlags.DirectionRightToLeft)
            MyBase.PrintHeader(graph)
            If Not View.OptionsPrint.PrintHeader Then
                Return
            End If
            Dim indent As New Point(Me.Indent, HeaderY)
            Dim r As Rectangle = Rectangle.Empty
            Dim usePrintStyles As Boolean = View.OptionsPrint.UsePrintStyles
            SetDefaultBrickStyle(graph, Bricks("HeaderPanel"))

            Dim HeaderHeight As Integer = (TryCast(Me.View, CustomBandedGridView)).HeaderHeightOnExport

            Dim columnBricks As New Dictionary(Of PrintColumnInfo, ITextBrick)()
            For Each col As PrintColumnInfo In Columns
                If Not usePrintStyles Then
                    Dim temp As New AppearanceObject()
                    AppearanceHelper.Combine(temp, New AppearanceObject() { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel })
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1))
                End If
                r = col.Bounds
                r = New Rectangle(r.Location, New Size(r.Width, HeaderHeight))
                r.Offset(indent)
                Dim mi As MethodInfo = GetType(GridColumn).GetMethod("GetTextCaptionForPrinting", BindingFlags.NonPublic Or BindingFlags.Instance)
                Dim caption As String = DirectCast(mi.Invoke(col.Column, New Object() { }), String) 'col.Column.GetTextCaptionForPrinting();
                If Not col.Column.OptionsColumn.ShowCaption Then
                    caption = String.Empty
                End If
                Dim itb As ITextBrick = DrawTextBrick(graph, caption, r)
                If (TypeOf col.Column Is CustomBandedColumn) AndAlso ((TryCast(col.Column, CustomBandedColumn)).CaptionVerticalOnExport) Then
                    itb.Style = vert
                Else
                    SetDefaultBrickStyle(graph, Bricks("HeaderPanel"))
                End If
                If caption.Contains(Environment.NewLine) Then
                    itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, True)
                End If

                If AppearancePrint.HeaderPanel.TextOptions.WordWrap = WordWrap.NoWrap AndAlso View.OptionsPrint.UsePrintStyles Then
                    Using g As Graphics = Me.View.GridControl.CreateGraphics()
                        Dim s As SizeF = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value)
                        If s.Width + 5 >= r.Width Then
                            itb.Text = ""
                            itb.TextValue = ""
                        End If
                    End Using
                End If
                columnBricks.Add(col, itb)
                Using g As Graphics = Me.View.GridControl.CreateGraphics()
                    Dim s As SizeF = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value)
                    If s.Height + 5 >= r.Height Then
                        itb.Text = ""
                        itb.TextValue = ""
                    End If
                End Using
            Next col
        End Sub
    End Class

    Friend Class MyBandedColumnCollection
        Inherits BandedGridColumnCollection

        Public Sub New(ByVal view As ColumnView)
            MyBase.New(view)

        End Sub
        Protected Overrides Function CreateColumn() As GridColumn
            Return New CustomBandedColumn()
        End Function

    End Class
    Friend Class CustomBandedColumn
        Inherits BandedGridColumn

        Public Sub New()
            CaptionVerticalOnExport = False
        End Sub
        Protected Overrides Sub Assign(ByVal column As GridColumn)
            If TypeOf column Is CustomBandedColumn Then
                Me.CaptionVerticalOnExport = CType(column, CustomBandedColumn).CaptionVerticalOnExport
            End If
            MyBase.Assign(column)
        End Sub
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)> _
        Public Property CaptionVerticalOnExport() As Boolean
    End Class
End Namespace
